runtime/buffer.h

```diff
@@ -192,6 +192,57 @@ class raw_buffer {
     }
   }
 
+  // Remove dimension `d` and move the base pointer to point to `at` in this dimension.
+  // `at` is dim(d).min() by default.
+  // If `d` is 0 or rank - 1, the slice does not mutate the dims array.
+  raw_buffer& slice(std::size_t d) {
+    assert(d < rank);
+    if (d == 0) {
+      dims += 1;
+    } else {
+      // Erase the sliced dim from the list.
+      std::move(dims + d + 1, dims + rank, dims + d);
+    }
+    rank -= 1;
+    return *this;
+  }
+  raw_buffer& slice(std::size_t d, index_t at) {
+    base = offset_bytes(base, dim(d).flat_offset_bytes(at));
+    return slice(d);
+  }
+
+  // Insert a new dimension `dim` at index d, increasing the rank by 1. This assumes that the buffer has previously been
+  // sliced using `slice` above, and will grow the dims array accordingly.
+  raw_buffer& unslice(std::size_t d, const slinky::dim& dim) {
```

Comment:
This one scares me -- what happens if I call unslice() by mistake on a never-sliced buffer?

Reply:
It's going to crash. This is the main reason I haven't moved this functionality to a helper function so far. There's just no good way to implement this, because we can't own the `dims` array to permit reallocating it as needed.

Reply:
I moved this to buffer<T>, where we can assert we own the dims and the new dimension fits. This makes it a lot less useful, we can't use it in evaluate for example. But I don't think I need this at the moment.